spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/MappingSpannerReadConverter.java

```diff
@@ -73,163 +78,201 @@
-						if (!allowMissingColumns) {
-							throw new SpannerDataException(
-									"Unable to read column from Spanner results: "
-											+ columnName,
-									e);
-						}
-					}
-					Class propType = spannerPersistentProperty.getType();
-
-					boolean valueSet;
-
-					/*
-					 * Due to type erasure, binder methods for Iterable properties must be
-					 * manually specified. ByteArray must be excluded since it implements
-					 * Iterable, but is also explicitly supported by spanner.
-					 */
-					if (ConversionUtils.isIterableNonByteArrayType(propType)) {
-						valueSet = attemptReadIterableValue(spannerPersistentProperty,
-								source, columnName, accessor);
-					}
-					else {
-						Class sourceType = ConversionUtils.SPANNER_COLUMN_TYPE_TO_JAVA_TYPE_MAPPING
-								.get(source.getColumnType(columnName));
-							valueSet = attemptReadSingleItemValue(
-									spannerPersistentProperty, source, sourceType,
-									columnName, accessor);
-					}
-
-					if (!valueSet) {
-						throw new SpannerDataException(String.format(
-								"The value in column with name %s"
-										+ " could not be converted to the corresponding property in the entity."
-										+ " The property's type is %s.",
-								columnName, propType));
+					if (!shouldSkipProperty(
+							structAccessor,
+							spannerPersistentProperty,
+							includeColumns,
+							readAllColumns,
+							persistenceConstructor,
+							allowMissingColumns)) {
+						attemptToReadProperty(structAccessor, spannerPersistentProperty, accessor);
 					}
 
 				});
 		return object;
 	}
 
-	@Override
-	public <R> R read(Class<R> type, Struct source) {
-		return read(type, source, null, false);
+	private void attemptToReadProperty(StructAccessor structAccessor,
+			SpannerPersistentProperty spannerPersistentProperty,
+			PersistentPropertyAccessor accessor) {
+		String columnName = spannerPersistentProperty.getColumnName();
+		Object value = structAccessor.get(spannerPersistentProperty);
+		if (value != null) {
+			accessor.setProperty(spannerPersistentProperty, value);
+		}
+		else {
+			throw new SpannerDataException(String.format(
+					"The value in column with name %s"
+							+ " could not be converted to the corresponding property in the entity."
+							+ " The property's type is %s.",
+					columnName, spannerPersistentProperty.getType()));
+		}
 	}
 
-	private boolean attemptReadSingleItemValue(
-			SpannerPersistentProperty spannerPersistentProperty, Struct struct,
-			Class sourceType,
-			String colName, PersistentPropertyAccessor accessor) {
-		Class targetType = spannerPersistentProperty.getType();
-		if (sourceType == null || !canConvert(sourceType, targetType)) {
-			return false;
-		}
-		BiFunction readFunction = singleItemReadMethodMapping
-				.get(ConversionUtils.boxIfNeeded(sourceType));
-		if (readFunction == null) {
-			return false;
+	private <R> boolean shouldSkipProperty(StructAccessor structAccessor,
```

Comment:
This type parameter `<R>` isn't necessary actually. The reason is that the only place it is used is as a constraint in an argument. This is saying "give me a PreferredConstruct, whose first type param must extend R, but I never say what R should be and the only thing telling me what R is the PreferredConstructor itself". As  a result it's not actually a restriction at all.

For example, if instead you had this method signature as `private <R,S> boolean shouldSkipProperty`, 
and the argument as `PreferredConstructor<? extends S, SpannerPersistentProperty> persistenceConstructor` everything still compiles and works despite there being total ambiguity about what R and S are.

Reply:
Sounds good.